File: packages/core/src/pipeline.js
import { link } from "./stages/link.js";
import { scan } from "./stages/scan.js";
import { dedupeHash, dedupeResolve } from "./stages/dedupe.js";
import { reconcile } from "./stages/reconcile.js";
import { rename } from "./stages/rename.js";
import { apply } from "./stages/apply.js";
import { report } from "./core/report.js";

export async function runPipeline(options, hooks = {}) {
  // Phase 1: scan files
  const rawCollections = await scan(options.source, options);
  hooks.onScan?.(rawCollections, options.source);

  // Phase 2: link media to sidecar files
  const matches = link(rawCollections, options);
  hooks.onLink?.(matches);

  const { manifest } = matches;
  if (!options.skipDedupe) {
    // Phase 3: calculate hash values for media files
    await dedupeHash(manifest, options);

    // Phase 4: use hashes to dedupe files
    await dedupeResolve(manifest, options);
    hooks.onDedupe?.(manifest);
  }

  if (!options.skipEmbed || !options.skipRename) {
    // Phase 5: resolve disk and sidecar metadata
    await reconcile(manifest, options);
    hooks.onReconcile?.(manifest);
  }

  if (!options.skipRename) {
    // Phase 6: compute target paths for renaming
    rename(manifest, options);
    hooks.onRename?.(manifest);
  }

  if (!options.dryRun) {
    // Phase 7: apply changes to disk
    await apply(manifest, options);
    hooks.onApply?.(manifest);

    // Write per-file CSV report to destination
    await report(manifest, options);
  }
}

File: packages/core/src/pixelkasten/pipeline.py
"""
Unified pipeline orchestrator.

Linear flow:
  scan → link → dedupe → reconcile → group → emit

Link matches sidecars automatically when present. emit writes GUID-named
copies and per-asset sidecars into <destination>/.pixelkasten/.
"""

from collections.abc import Callable

from pixelkasten.configuration import Hooks, Options
from pixelkasten.manifest import ManifestEntry
from pixelkasten.stages.report import report
from pixelkasten.stages.emit import emit
from pixelkasten.stages.dedupe import dedupe_hash, dedupe_resolve
from pixelkasten.stages.group import group
from pixelkasten.stages.link import link
from pixelkasten.stages.reconcile import reconcile
from pixelkasten.stages.scan import scan


def run_pipeline(options: Options, hooks: Hooks, progress: Callable) -> list[ManifestEntry]:
    """
    Unified pipeline: linear flow of stages.

    Returns the manifest after all stages have run.
    """

    raw_collections = scan(options.source)
    hooks.on_scan(raw_collections, options.source)

    result = link(raw_collections, options)
    hooks.on_link(result)

    manifest = result["manifest"]

    if not options.skip_dedupe:
        with progress("Hashing files", len(manifest)) as tick:
            dedupe_hash(manifest, on_progress=tick)
        dedupe_resolve(manifest, options)
        hooks.on_dedupe(manifest)

    count = sum(1 for e in manifest if e.can_keep())
    with progress("Reading metadata", count) as tick:
        reconcile(manifest, options, on_progress=tick)
    hooks.on_reconcile(manifest)

    group(manifest, options)

    if not options.dry_run:
        count = sum(1 for e in manifest if e.can_keep())
        with progress("Emitting working library", count) as tick:
            emit(manifest, options, on_progress=tick)
        hooks.on_apply(manifest)
        report(manifest, options)

    hooks.on_errors(manifest)

    if options.write_manifest:
        from pixelkasten.tools.serialize import write_manifest

        write_manifest(manifest, options)

    return manifest
